src/nonLinPMMH.cpp

```diff
@@ -0,0 +1,146 @@
+// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+//
+// nonLinPMMH.cpp: Example 3.1 of Andrieu et al. (2010). Implementing particle marginal
+// Metropolis-Hastings for a toy non-linear state space model previously described in
+// Gordon et al. (1993) and Kitagawa (1996).
+//
+// Copyright (C) 2017         Dirk Eddelbuettel, Adam Johansen and Leah South
+//
+// This file is part of RcppSMC.
+//
+// RcppSMC is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// RcppSMC is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with RcppSMC.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "nonLinPMMH.h"
+
+namespace nonLinPMMH {
+    const double a_prior = 0.01;
+	const double b_prior = 0.01;
+}
+
+using namespace std;
+using namespace nonLinPMMH;
+
+
+// nonLinPMMH_impl() function callable from R via Rcpp::
+// [[Rcpp::export]]
+Rcpp::DataFrame nonLinPMMH_impl(arma::vec data, unsigned long lNumber, unsigned long lMCMCits) { 	
+  
+  try {
+	arma::vec sigv(lMCMCits), sigw(lMCMCits);
+	long lIterates = data.n_rows;
+    y = data;
+	
+	arma::vec loglike = arma::zeros(lMCMCits);
+	arma::vec logprior = arma::zeros(lMCMCits);
+	
+	double dRand;
+	double loglike_prop;
+	double logprior_prop;
+	
+    //Initialise and run the sampler
+    smc::sampler<double> Sampler(lNumber, HistoryType::NONE);
+	theta_prop.sigv = 10.0;
+	theta_prop.sigw = 10.0;
+		
+	sigv(0) = theta_prop.sigv;
+	sigw(0) = theta_prop.sigw;
+	
+	// Getting a particle filtering estimate of the log likelihood.
+	smc::moveset<double> Moveset(fInitialise, fMove, NULL);
+    Sampler.SetResampleParams(ResampleType::MULTINOMIAL, 0.5);
+    Sampler.SetMoveSet(Moveset);
+    Sampler.Initialise();
+	Sampler.IterateUntil(lIterates-1);
+	loglike(0) = Sampler.GetLogNCPath();
+	
+	// Inverse gamma prior
+	logprior(0) = logPrior(theta_prop);
+    
+	double MH_ratio;
+	for (unsigned int i = 1; i<lMCMCits; i++){
+		// RW proposal for parameters
+		theta_prop.sigv = R::rnorm(sigv(i-1),0.15);
```

Comment:
Thanks, that's a good point. I've used vectors for this now. I tried using profvis quickly to compare the two approaches. I did 100 iterations without any plotting and the non-vectorised time was 91040 (milliseconds I think) and vectorised was 92240, so it was slightly slower with vectorisation. That's probably not a very fair test though -- 100 iterations is unrealistically small.

I can do a proper check with a more realistic number of iterations and look at time and memory if that's useful.